CI: wpt upstreamer fails when git process output contains invalid UTF-8

Out of curiosity, I tried running the new upstreamer (#29601) against #29594 with a merge from master, which is currently more challenging for the script because it applies the commits one at a time.

https://github.com/servo/servo/actions/runs/4675613330/jobs/8280932770?pr=29594

As expected, the script processed a lot of WPT changes made outside my PR, but it failed on them in an interesting way:

```
INFO:root:  → Execution (cwd='./servo'): git diff b47513006044f769c6c34952028dc163de77982c b47513006044f769c6c34952028dc163de77982c~6 -- tests/wpt/web-platform-tests/
ERROR:root:'utf-8' codec can't decode byte 0xa2 in position 687773: invalid start byte
Traceback (most recent call last):
  File "/home/runner/work/servo/servo/servo/etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py", line 182, in run
    self.handle_new_pull_request_contents(run, pull_data)
  File "/home/runner/work/servo/servo/servo/etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py", line 202, in handle_new_pull_request_contents
    self.local_servo_repo.run(
  File "/home/runner/work/servo/servo/servo/etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py", line 66, in run
    ).decode("utf-8")
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa2 in position 687773: invalid start byte
Error: Process completed with exit code 1.
```

This is because LocalGitRepo.run assumes that the output can be decoded as UTF-8, but git-diff(1) and git-show(1) output can contain arbitrary high bytes unless git [gives up and decides the file is binary](https://stackoverflow.com/a/6134127).

We should leave git-diff(1) and git-show(1) output as bytes where possible, and if we really need to process them as str (for example, to find `---` and `+++`), we should consider using [surrogateescape](https://docs.python.org/3/library/codecs.html#error-handlers) to treat it as ASCII with opaque high bytes.

We should also check that we handle weird paths correctly, even though Windows and macOS support means we won’t need to deal with any invalid UTF-8. Most git commands like `git status` and `git diff` will use a special double-quote-octal escaping for paths with *any* high bytes (or double quotes), whereas `-z` modes like `git status -z` and `git diff --numstat -z` will happily spit out arbitrary high bytes.

```
% touch foo bér $'\xFF' '"'
% git add foo bér $'\xFF' '"'
% git status -s    
A  "\""
A  "b\303\251r"
A  foo
A  "\377"

% git diff --cached | rg '^diff '
diff --git "a/\"" "b/\""
diff --git "a/b\303\251r" "b/b\303\251r"
diff --git a/foo b/foo
diff --git "a/\377" "b/\377"

% git status -z | tr \\0 \\n
A  "
A  bér
A  foo
A  �

% git diff --cached --numstat -z | tr \\0 \\n
0	0	"
0	0	bér
0	0	foo
0	0	�
```

## more details

Inspecting the output with [utf8check](https://github.com/delan/utf8check):

```
% git diff b47513006044f769c6c34952028dc163de77982c b47513006044f769c6c34952028dc163de77982c~6 -- tests/wpt/web-platform-tests/ | ~/code/utf8check/utf8check             
688567: unexpected continuation byte; ASCII or start byte expected
688779: unexpected continuation byte; ASCII or start byte expected
691366: unexpected continuation byte; ASCII or start byte expected
```
```
% git diff b47513006044f769c6c34952028dc163de77982c b47513006044f769c6c34952028dc163de77982c~6 -- tests/wpt/web-platform-tests/ | dd status=none bs=1 skip=688000 | less
% git diff b47513006044f769c6c34952028dc163de77982c b47513006044f769c6c34952028dc163de77982c~6 -- tests/wpt/web-platform-tests/ | dd status=none bs=1 skip=691000 | less
(finding the offending paths)
```
```
% git log -- tests/wpt/web-platform-tests/html/syntax/charset/inheritance-bogus-meta.html tests/wpt/web-platform-tests/html/syntax/charset/resources/bogus-charset.html | rg '^commit '
commit bb34f95b33cd3b919fda221408720e7e6dea84ab
```
```
% git show bb34f95b33cd3b919fda221408720e7e6dea84ab tests/wpt/web-platform-tests/html/syntax/charset/inheritance-bogus-meta.html tests/wpt/web-platform-tests/html/syntax/charset/resources/bogus-charset.html | ~/code/utf8check/utf8check -s
commit bb34f95b33cd3b919fda221408720e7e6dea84ab
Author: WPT Sync Bot <josh+wptsync@joshmatthews.net>
Date:   Fri Apr 7 01:27:34 2023 +0000

    Update web-platform-tests to revision b'1d9b01e2fad6af3a057d571b1e088e15fa9bc8e6'

diff --git a/tests/wpt/web-platform-tests/html/syntax/charset/inheritance-bogus-meta.html b/tests/wpt/web-platform-tests/html/syntax/charset/inheritance-bogus-meta.html
new file mode 100644
index 0000000000..616ba51853
--- /dev/null
+++ b/tests/wpt/web-platform-tests/html/syntax/charset/inheritance-bogus-meta.html
@@ -0,0 +1,46 @@
+<!doctype html>
+<meta charset=windows-1253>
+<title>Inheriting from windows-1253 parent</title>
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+<script src=/common/get-host-info.sub.js></script>
+<div id=log></div>
+<script>
+[
+  {
+    "title": "Child with bogus <meta charset>",
+    "url": "resources/bogus-charset.html",
+    "expected": "�\n" // 0x00A2 in windows-1252 is at the same position as 0x0386 in windows-1253
+  },
+  {
+    "title": "Child with bogus Content-Type charset",
+    "url": "resources/bogus-charset-http.py",
+    "expected": "�\n"
+  },
+  {
+    "title": "Child with bogus Content-Type charset, but valid <meta charset>",
+    "url": "resources/bogus-charset-http-valid-meta.py",
+    "expected": "\u045E\n"
+  }
+].forEach(({ title, url, expected }) => {
+  async_test(t => {
+    const frame = document.createElement("iframe");
+    t.add_cleanup(() => frame.remove());
+    frame.src = url;
+    frame.onload = t.step_func_done(() => {
+      assert_equals(frame.contentDocument.body.textContent, expected);
+    });
+    document.body.append(frame);
+  }, title);
+});
+
+async_test(t => {
+  self.onmessage = t.step_func_done(({ data }) => {
+    assert_equals(data, "\u00A2\n");
+  });
+  const frame = document.createElement("iframe");
+  t.add_cleanup(() => frame.remove());
+  frame.src = get_host_info().HTTP_REMOTE_ORIGIN + new URL("resources/bogus-charset.html", location).pathname + "?postMessage";
+  document.body.append(frame);
+}, "Cross-origin child with bogus <meta charset>");
+</script>
diff --git a/tests/wpt/web-platform-tests/html/syntax/charset/resources/bogus-charset.html b/tests/wpt/web-platform-tests/html/syntax/charset/resources/bogus-charset.html
new file mode 100644
index 0000000000..e77e874466
--- /dev/null
+++ b/tests/wpt/web-platform-tests/html/syntax/charset/resources/bogus-charset.html
@@ -0,0 +1,7 @@
+<!doctype html><meta charset=this-is-not-a-charset><script>
+onload = () => {
+  if (location.search === "?postMessage") {
+    parent.postMessage(document.body.textContent, "*");
+  }
+}
+</script>�
```

The three � characters are A2h bytes, which are valid in Windows-1252 but not UTF-8.

--- a/etc/ci/upstream-wpt-changes/test.py
+++ b/etc/ci/upstream-wpt-changes/test.py
@@ -229,7 +229,7 @@ class TestApplyCommitsToWPT(unittest.TestCase):
 
     def run_test(self, pr_number: int, commit_data: dict):
         def make_commit(data):
-            with open(os.path.join(TESTS_DIR, data[2]), encoding="utf-8") as file:
+            with open(os.path.join(TESTS_DIR, data[2]), "rb") as file:
                 return {"author": data[0], "message": data[1], "diff": file.read()}
 
         commits = [make_commit(data) for data in commit_data]
@@ -272,6 +272,15 @@ class TestApplyCommitsToWPT(unittest.TestCase):
             [
                 ["test author <test@author>", "test commit message", "18746.diff"],
                 ["another person <two@author>", "a different message", "wpt.diff"],
+                ["another person <two@author>", "adding some non-utf8 chaos", "add-non-utf8-file.diff"],
+            ],
+        )
+
+    def test_non_utf8_commit(self):
+        self.run_test(
+            100,
+            [
+                ["test author <nonutf8@author>", "adding some non-utf8 chaos", "add-non-utf8-file.diff"],
             ],
         )
 
@@ -453,6 +462,17 @@ class TestFullSyncRun(unittest.TestCase):
                 "CommentStep:servo/servo#18746:🤖 This change no longer contains upstreamable changes "
                 "to WPT; closed existing upstream pull request (wpt/wpt#1).",
             ]
+        )
+
+    def test_opened_upstreamable_pr_with_non_utf8_file_contents(self):
+        self.assertListEqual(
+            self.run_test("opened.json", ["add-non-utf8-file.diff"]),
+            [
+                "CreateOrUpdateBranchForPRStep:1:servo-wpt-sync/wpt/servo_export_18746",
+                "OpenPRStep:servo-wpt-sync/wpt/servo_export_18746→wpt/wpt#1",
+                "CommentStep:servo/servo#18746:🤖 Opened new upstream WPT pull request "
+                "(wpt/wpt#1) with upstreamable changes.",
+            ],
         )
 
     def test_open_new_upstreamable_pr_with_preexisting_upstream_pr_not_apply_cleanly_to_upstream(

--- a/etc/ci/upstream-wpt-changes/tests/add-non-utf8-file.diff
+++ b/etc/ci/upstream-wpt-changes/tests/add-non-utf8-file.diff
@@ -1,0 +1,7 @@
+diff --git a/tests/wpt/web-platform-tests/non-utf8-file.txt b/tests/wpt/web-platform-tests/non-utf8-file.txt
+new file mode 100644
+index 0000000..5a992e0
+--- /dev/null
++++ b/tests/wpt/web-platform-tests/non-utf8-file.txt
+@@ -0,0 +1 @@
++foo bÃ©r ÿ "

--- a/etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py
+++ b/etc/ci/upstream-wpt-changes/wptupstreamer/__init__.py
@@ -50,7 +50,7 @@ class LocalGitRepo:
         self.path = path
         self.sync = sync
 
-    def run(self, *args, env: dict = {}):
+    def run_without_encoding(self, *args, env: dict = {}):
         command_line = ["git"] + list(args)
         logging.info("  → Execution (cwd='%s'): %s",
                      self.path, " ".join(command_line))
@@ -63,11 +63,18 @@ class LocalGitRepo:
         try:
             return subprocess.check_output(
                 command_line, cwd=self.path, env=env, stderr=subprocess.STDOUT
-            ).decode("utf-8")
+            )
         except subprocess.CalledProcessError as exception:
             logging.warning("Process execution failed with output:\n%s",
-                            exception.output.decode("utf-8"))
+                            exception.output.decode("utf-8", errors="surrogateescape"))
             raise exception
+
+    def run(self, *args, env: dict = {}):
+        return (
+            self
+            .run_without_encoding(*args, env=env)
+            .decode("utf-8", errors="surrogateescape")
+        )
 
 
 @dataclasses.dataclass()

--- a/etc/ci/upstream-wpt-changes/wptupstreamer/step.py
+++ b/etc/ci/upstream-wpt-changes/wptupstreamer/step.py
@@ -114,7 +114,9 @@ class CreateOrUpdateBranchForPRStep(Step):
             # TODO: If we could cleverly parse and manipulate the full commit diff
             # we could avoid cloning the servo repository altogether and only
             # have to fetch the commit diffs from GitHub.
-            diff = local_servo_repo.run(
+            # NB: The output of git show might include binary files or non-UTF8 text,
+            # so store the content of the diff as a `bytes`.
+            diff = local_servo_repo.run_without_encoding(
                 "show", "--binary", "--format=%b", sha, "--", UPSTREAMABLE_PATH
             )
 
@@ -140,7 +142,7 @@ class CreateOrUpdateBranchForPRStep(Step):
         strip_count = UPSTREAMABLE_PATH.count("/") + 1
 
         try:
-            with open(patch_path, "w", encoding="utf-8") as file:
+            with open(patch_path, "wb") as file:
                 file.write(commit["diff"])
             run.sync.local_wpt_repo.run(
                 "apply", PATCH_FILE_NAME, "-p", str(strip_count)

